jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/bpmn2/BPMN2DataServiceImpl.java

```diff
@@ -131,9 +83,29 @@ private void validateNonEmptyDeploymentIdAndProcessIdAndTaskName(String deployme
             throw new IllegalStateException( msg.toString() );
         }
     }
+    
+    public void addProcessDefinition(String deploymentId, String processId, Object processDescriptor, KieContainer kieContainer) {
+        Map<String, ProcessDescriptor> definitions = null;
+        synchronized (definitionCache) {
+            definitions = definitionCache.get(deploymentId);
+            if (definitions == null) {
```

Comment:
Could you use putIfAbsent() here instead of a null check? The null check will lead to a race condition.